src/lib/cimb_translator/FloodDecodePositions.cpp

```diff
@@ -47,30 +55,81 @@ FloodDecodePositions::iter FloodDecodePositions::next()
 
 		std::vector<bool>::reference needsDecode = _remaining[i];
 		if (!needsDecode)
-		    continue;
+			continue;
 
 		needsDecode = false;
 		++_count;
-		auto [drift, __] = _instructions[i];
-		return {i, _positions[i], drift};
+		auto [drift, __, cooldown] = _instructions[i];
+		return {i, _positions[i], drift, cooldown};
 	}
 
-	return {0, {0, 0}, CellDrift()};
+	return {0, {0, 0}, CellDrift(), 0xFF};
 }
 
-int FloodDecodePositions::update(unsigned index, const CellDrift& drift, unsigned error_distance)
+int FloodDecodePositions::update_adjacents(const std::array<int,4>& adj, const CellDrift& drift, unsigned error_distance, uint8_t cooldown)
 {
-	std::array<int,4> adj = _cellFinder.find(index);
 	for (int next : adj)
 	{
 		if (next < 0 or !_remaining[next])
 			continue;
 		decode_instructions& di = _instructions[next];
-		if (di.second <= error_distance)
+		if (std::get<1>(di) <= error_distance)
 			continue;
-		di = {drift, error_distance};
+		di = {drift, error_distance, cooldown};
 		_heap.push({next, error_distance});
 	}
+
+	return 0;
+}
+
+int FloodDecodePositions::update(unsigned index, const CellDrift& drift, unsigned error_distance, uint8_t cooldown)
+{
+	std::array<int,4> adj = _cellFinder.find(index);
+	update_adjacents(adj, drift, error_distance, cooldown);
+
+	auto& [_, prev_error, prev_cooldown] = _instructions[index];
+	// in the case where we have consecutive high confidence cells with no drift changes,
+	// it's safe(ish) to aggressively queue a few more cells
+	if (prev_error < 3 and error_distance < 3 and prev_cooldown == 4 and cooldown == 4)
+	{
+		unsigned rr = 0;
+		unsigned ll = 1;
+		unsigned dd = 2;
+		unsigned uu = 3;
+
+		int rridx = adj[rr];
+		int llidx = adj[ll];
+		if (rridx >= 0 and llidx >= 0)
+		{
+			std::array<int,4> horizon = {-1, -1, -1, -1};
```

Comment:
The reason this is array<4> is simply convenience. That's the size we expect for the directional calculation (up,down,left,right), so it's reused here.